File: services/veilkey-cli/project_config.go
package main

import (
	"os"
	"path/filepath"

	"gopkg.in/yaml.v3"
)

type ProjectConfig struct {
	PatternsFile string   `yaml:"patterns"`
	ExcludePaths []string `yaml:"exclude_paths"`
	Format       string   `yaml:"format"`
	APIURL       string   `yaml:"api_url"`
	ExitCode     bool     `yaml:"exit_code"`
}

// LoadProjectConfig loads a .veilkey.yml project config.
// Search order: explicit path → .veilkey.yml in CWD → .veilkey.yaml in CWD → ~/.config/veilkey/config.yml.
// Returns (nil, nil) if no config file is found.
func LoadProjectConfig(path string) (*ProjectConfig, error) {
	if path != "" {
		return loadProjectConfigFile(path)
	}

	candidates := []string{
		".veilkey.yml",
		".veilkey.yaml",
	}

	// CWD candidates
	for _, name := range candidates {
		if _, err := os.Stat(name); err == nil {
			return loadProjectConfigFile(name)
		}
	}

	// User-level fallback
	home, err := os.UserHomeDir()
	if err == nil {
		userConfig := filepath.Join(home, ".config", "veilkey", "config.yml")
		if _, err := os.Stat(userConfig); err == nil {
			return loadProjectConfigFile(userConfig)
		}
	}

	return nil, nil
}

func loadProjectConfigFile(path string) (*ProjectConfig, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var cfg ProjectConfig
	if err := yaml.Unmarshal(data, &cfg); err != nil {
		return nil, err
	}

	return &cfg, nil
}

File: services/veil-cli/src/project_config.rs
use serde::Deserialize;
use std::fs;
use std::path::PathBuf;

#[derive(Debug, Deserialize, Default)]
#[allow(dead_code)]
pub struct ProjectConfig {
    #[serde(rename = "patterns", default)]
    pub patterns_file: String,
    #[serde(default)]
    pub exclude_paths: Vec<String>,
    #[serde(default)]
    pub format: String,
    #[serde(default)]
    pub api_url: String,
    #[serde(default)]
    pub exit_code: bool,
}

/// Search order: explicit path → $VEILKEY_CONFIG → .veilkey.yml → .veilkey.yaml → ~/.config/veilkey/config.yml
/// Returns None if no config file is found or parse fails.
pub fn load_project_config(path: Option<&str>) -> Option<ProjectConfig> {
    if let Some(p) = path {
        return load_file(p);
    }

    if let Ok(p) = std::env::var("VEILKEY_CONFIG") {
        if !p.is_empty() {
            return load_file(&p);
        }
    }

    for name in &[".veilkey.yml", ".veilkey.yaml"] {
        if std::path::Path::new(name).exists() {
            return load_file(name);
        }
    }

    if let Ok(home) = std::env::var("HOME") {
        let user_cfg = PathBuf::from(home)
            .join(".config")
            .join("veilkey")
            .join("config.yml");
        if user_cfg.exists() {
            return load_file(&user_cfg.to_string_lossy());
        }
    }

    None
}

fn load_file(path: &str) -> Option<ProjectConfig> {
    let data = fs::read_to_string(path).ok()?;
    serde_yaml::from_str(&data).ok()
}
